File: frontend/lib/models/chat_model.dart
import 'package:frontend/models/message_model.dart';

class ChatModel {
  const ChatModel({
    required this.chatId,
    required this.taskId,
    required this.taskTitle,
    required this.taskOwnerUserId,
    required this.taskOwnerName,
    required this.users,
    required this.isClosed,
    required this.lastMessage,
  });

  final String chatId;
  final String taskId;
  final String taskTitle;
  final String taskOwnerUserId;
  final String taskOwnerName;
  final List<String> users;
  final bool isClosed;
  final MessageModel lastMessage;

  factory ChatModel.fromJson(Map<String, dynamic> json) {
    return ChatModel(
      chatId: json['chatId'] as String,
      taskId: json['taskId'] as String,
      taskTitle: json['taskTitle'] as String,
      taskOwnerUserId: json['taskOwnerUserId'] as String,
      taskOwnerName: json['taskOwnerName'] as String,
      users: (json['users'] as List<dynamic>).cast<String>(),
      isClosed: json['isClosed'] as bool? ?? false,
      lastMessage: MessageModel.fromJson(
        json['lastMessage'] as Map<String, dynamic>,
      ),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'chatId': chatId,
      'taskId': taskId,
      'taskTitle': taskTitle,
      'taskOwnerUserId': taskOwnerUserId,
      'taskOwnerName': taskOwnerName,
      'users': users,
      'isClosed': isClosed,
      'lastMessage': lastMessage.toJson(),
    };
  }
}

File: backend-nodejs/src/models/chat.model.js
const {toMessageRecord} = require('./message.model');

function normalizeString(value) {
	return typeof value === 'string' ? value.trim() : '';
}

function createFallbackLastMessage(chat = {}) {
	const chatId = normalizeString(chat.chatId);
	const taskId = normalizeString(chat.taskId);

	return {
		id: '',
		chatId,
		taskId,
		senderId: '',
		text: '',
		timestamp: new Date().toISOString(),
		isPaymentRequest: false,
	};
}

function toChatRecord(chat) {
	if (!chat || typeof chat !== 'object') {
		return null;
	}

	return {
		chatId: normalizeString(chat.chatId),
		taskId: normalizeString(chat.taskId),
		taskTitle: normalizeString(chat.taskTitle),
		taskOwnerUserId: normalizeString(chat.taskOwnerUserId),
		taskOwnerName: normalizeString(chat.taskOwnerName),
		users: Array.isArray(chat.users) ? chat.users.filter((entry) => typeof entry === 'string') : [],
		isClosed: Boolean(chat.isClosed),
		lastMessage: toMessageRecord(chat.lastMessage) || createFallbackLastMessage(chat),
	};
}

function isValidChatRecord(chat) {
	return Boolean(
		chat &&
			typeof chat.chatId === 'string' &&
			typeof chat.taskId === 'string' &&
			typeof chat.taskTitle === 'string' &&
			typeof chat.taskOwnerUserId === 'string' &&
			typeof chat.taskOwnerName === 'string' &&
			Array.isArray(chat.users),
	);
}

module.exports = {
	normalizeString,
	toChatRecord,
	isValidChatRecord,
};
